java/arcs/core/analysis/ClaimDeduction.kt

```diff
@@ -0,0 +1,123 @@
+package arcs.core.analysis
+
+import arcs.core.data.AccessPath
+import arcs.core.data.Claim
+import arcs.core.data.HandleConnectionSpec
+import arcs.core.data.ParticleSpec
+import arcs.core.data.Recipe
+import arcs.core.data.expression.Expression
+
+/**
+ * Deduce [Claim]s on Paxel [Expression]s from all connections on a [Recipe.Particle].
+ *
+ * See [DependencyNode] for more information.
+ */
+fun RecipeGraph.Node.Particle.deduceClaims(): List<Claim> {
+  return particle.spec.connections.values
+    .filter { connection -> connection.expression != null }
+    .flatMap { connection ->
+      val root = connection.expression?.analyze()
+      require(root is DependencyNode.AssociationNode) {
+        "Expression on '${particle.spec.name}.${connection.name}' is invalid."
+      }
+      root.toClaims(particle, connection)
+    }
+}
+
+/**
+ * This flattens nested [DependencyNode]s into a map of [Path]s and terminal [DependencyNode]s.
+ *
+ * Example:
+ *   ```
+ *   DependencyNode.AssociationNode(
+ *     "a" to DependencyNode.AssociationNode(
+ *        "x" to DependencyNode.AssociationNode(
+ *          "q": DependencyNode.Input("input", "foo")
+ *        ),
+ *        "y" to DependencyNode.Input("input", "bar")
+ *     ),
+ *     "b" to DependencyNode.DerivedFrom(
+ *       DependencyNode.Input("input", "foo"),
+ *       DependencyNode.Input("input", "bar")
+ *     )
+ *   )
+ *   ```
+ * This would be flattened to:
+ *
+ *   ```
+ *   mapOf(
+ *     listOf("a", "x", "q") to DependencyNode.Input("input", "foo"),
+ *     listOf("a", "y") to DependencyNode.Input("input", "bar"),
+ *     listOf("b") to DependencyNode.DerivedFrom(
+ *       DependencyNode.Input("input", "foo"),
+ *       DependencyNode.Input("input", "bar")
+ *     )
+ *   )
+ *   ```
+ */
+private fun DependencyNode.AssociationNode.flatten(
+  prefix: List<String> = emptyList()
+): Map<List<String>, DependencyNode> {
+  val output = mutableMapOf<List<String>, DependencyNode>()
+  for ((key, value) in this.associations) {
+    val path = prefix + listOf(key)
+    when (value) {
+      is DependencyNode.AssociationNode -> output += value.flatten(path)
```

Comment:
It might be more efficient to simply pass `output` down to `flatten` and update the map directly in the `else` case.